### PR TITLE
Enable sentinel logging to discover pages where old code is run

### PR DESCRIPTION
**Pull request by marsavar**
## What does this change?
This change introduces a new module called `sentinel.ts` containing a function named `amIUsed`.

`amIUsed` can be placed anywhere in the `frontend` codebase and will trigger a Sentry logging event any time a piece of code containing `amIUsed` is executed. The Sentry events are labelled with the tag `commercial-sentinel`, allowing for straightforward filtering. 

Sentry logging in `frontend` relies on `lib/raven.js`, a wrapper around `raven-js`, a deprecated library that was replaced by `@sentry/browser` in 2019. Since `frontend` is due to be decommissioned in the near future, it made sense to reuse the existing `lib/raven` instead of writing code to configure the new SDK. However, this meant that changes to the sampling logic in `lib/raven` had to be made in order to accommodate for commercial logging: events labelled with the `commercial-sentinel` tag will now always bypass the default 0.08% sampling.

In order to avoid being potentially overwhelmed by Sentry notifications, a new commercial switch called `sentinelLogger` has been created, which will allows us to quickly disable logging, should the amount of notifications prove to be unmanageable.

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

## Screenshots

Logs can be filtered by tag on the search page in the Sentry UI ([here](https://sentry.io/organizations/the-guardian/issues/?project=35463&query=tag%3Acommercial-sentinel)). 

In the example below, we can see that the `init` function within the `article-aside-adverts` module runs on the UK business front.

image

## What is the value of this and can you measure success?
This will give us more confidence to delete unused modules within the `frontend` repository, which has the benefit of reducing the bundle size as well as helping us transition away from `frontend`.

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [x] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [x] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Does this change update the version of CAPI we're using?

- [x] No, all the existing database files are just fine
- [ ] Yes, and I have [re-run all the tests locally and checked in all the updated data/database/xyz files](https://github.com/guardian/frontend/blob/main/docs/03-dev-howtos/15-updating-test-database.md)

### Tested

- [x] Locally
- [x] On CODE (optional)

### CODE REVIEW

**Review by ashishpuliyel — APPROVED** [review]
Good job Mario! Neat way to end the week

File: static/src/javascripts/lib/raven.js [hunk 1]
```diff
@@ -12,6 +12,7 @@ const sentryOptions = {
         // localhost will not log errors, but call `shouldSendCallback`
         /localhost/,
         /assets\.guim\.co\.uk/,
+        /assets-code\.guim\.co\.uk/,
```
Comment (arelra) [inline static/src/javascripts/lib/raven.js:15]: Do we want to leave this in for CODE?

File: static/src/javascripts/lib/raven.js [hunk 2]
```diff
@@ -57,12 +58,18 @@ const sentryOptions = {
         const isIgnored =
             typeof data.tags.ignored !== 'undefined' && data.tags.ignored;
         const { enableSentryReporting } = config.get('switches');
-        const isInSample = Math.random() < 0.008; // 0.8%
+        const isSentinelLoggingEvent = data.tags.tag === 'commercial-sentinel'
```
Comment (arelra) [inline static/src/javascripts/lib/raven.js:61]: Wondering if its possible for `data.tags` to be `undefined` ? If so might be good to be defensive here with some optional chaining ie `data?.tags?.tag`

File: static/src/javascripts/lib/raven.js (hunk 1 above)
Comment (MarSavar) [reply static/src/javascripts/lib/raven.js:15]: Probably not! Will remove.

File: static/src/javascripts/lib/raven.js (hunk 2 above)
Comment (MarSavar) [reply static/src/javascripts/lib/raven.js:61]: You're right, `lib/raven` is used by other modules, not just `sentinel.ts`, so I think it's definitely possible for `data.tags` to be undefined.

File: static/src/javascripts/projects/commercial/sentinel.spec.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import { mocked } from 'ts-jest/utils';
+import raven_ from 'lib/raven';
```
Comment (mxdvl) [inline static/src/javascripts/projects/commercial/sentinel.spec.ts:2]: I don’t think you need to import raven or config, if you’re not using them explicitly and you have mocks for them.

File: static/src/javascripts/projects/commercial/sentinel.spec.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import config_ from '../../lib/config';
+import { amIUsed } from './sentinel';
+
+const raven = raven_;
+
+const config = config_ as {
+	get: (s: string, d: boolean) => boolean;
+};
```
Comment (mxdvl) [inline static/src/javascripts/projects/commercial/sentinel.spec.ts:10]: Ditto for importing mocks, you can ignore them entirely, I think.

```suggestion

import { amIUsed } from './sentinel';

```

File: static/src/javascripts/projects/commercial/sentinel.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+export const amIUsed = (
+	moduleName: string,
+	functionName: string,
+	label?: string,
+): void => {
+	// The function will return early if the sentinelLogger switch is disabled.
+	if (!config.get('switches.sentinelLogger', false)) return;
+
+	const functionToCheck = [moduleName, functionName, label]
+		.filter(Boolean)
```
Comment (mxdvl) [inline static/src/javascripts/projects/commercial/sentinel.ts:31]: Making use of the fact that `Boolean(undefined) == false`. Cheeky :)

**Review by mxdvl — APPROVED** [review]

File: static/src/javascripts/projects/commercial/sentinel.spec.ts (hunk 3 above)
Comment (arelra) [reply static/src/javascripts/projects/commercial/sentinel.spec.ts:2]: This is using the existing pattern elsewhere in the codebase for `mockReturnValue`.
We do need an explicit handle on the module later to enforce return values with `mockReturnValue`.
However I suspect you're right in that we might be able to use jest's auto mocking.
I'll have a look at simplifying and if possible raise another PR.

Comment (arelra) [reply static/src/javascripts/projects/commercial/sentinel.spec.ts:2]: Et voila: https://github.com/guardian/frontend/pull/23887

**Comment by mxdvl** [comment]
Since #23973, we’re using a different endpoint for `sentinel.ts` and I think the raven.js code can be removed.

**Pull request merged**